witness/golang/omniwitness/usbarmory/storage/slots/journal.go

```diff
@@ -0,0 +1,258 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package slots provides a simple "postbox" type filesystem.
+package slots
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+// magic0 is the only known journal header prefix.
+const magic0 = "TFJ0"
+
+// BlockReaderWriter describes a type which knows how to read and write
+// whole blocks to some backing storage.
+type BlockReaderWriter interface {
+	// BlockSize returns the block size of the underlying storage system.
+	BlockSize() uint
+
+	// ReadBlocks reads len(b) bytes into b from contiguous storage blocks starting
+	// at the given block address.
+	// b must be an integer multiple of the device's block size.
+	ReadBlocks(lba uint, b []byte) error
+
+	// WriteBlocks writes len(b) bytes from b to contiguous storage blocks starting
+	// at the given block address.
+	// b must be an integer multiple of the device's block size.
+	WriteBlocks(lba uint, b []byte) error
+}
+
+// Journal implements a record-based format which provides a resilient storage.
+type Journal struct {
+	dev           BlockReaderWriter
+	start         uint
+	length        uint
+	current       entry
+	nextBlock     uint
+	maxWriteBytes uint
+}
+
+// entry represents an entry in the journal.
+type entry struct {
+	// Magic is a 4 byte prefix which allows us to quickly filter out invalid entry records.
+	// This ie expected to contain the value in the magic0 const above.
+	Magic [4]byte
+	// Revision is an incrementing counter which tracks the total number of successful
+	// updates to a journal.
+	// Each successive entry should have a revision one greater than the previous entry.
+	Revision uint32
+	// DataLen is the length in bytes of the application data associated with this entry.
+	DataLen uint64
+	// DataSHA256 is the SHA256 hash of the application data associated with this entry.
+	DataSHA256 [32]byte
+	// Data is the application data associated with this entry.
+	Data []byte
+}
+
+// entryHeaderSize is the on-disk size of an entry without application data.
+const entryHeaderSize = 4 + 4 + 8 + 32
+
+// Size returns the number of bytes used by this entry record.
+func (e *entry) Size() int {
+	return entryHeaderSize + len(e.Data)
+}
+
+// OpenJournal returns a new journal structure for interacting with a journal stored in the
+// [start, start+length) range of blocks accessible via dev.
+func OpenJournal(dev BlockReaderWriter, start, length uint) (*Journal, error) {
```

Comment:
A wee discussion about the expectations around locking / ownership responsibilities would be very useful. I imagined that this would have a mutex in it, and the contract would be that there should be only one journal covering any particular range. However, given the lack of a mutex in here, either a) it's not thread-safe, or b) all locking and safety is performed at a higher level.